### ISSUE TITLE
Exception thrown from all_gather_async op with 2D tensor

### ISSUE THREAD

**Issue by hkwonTT** [tenstorrent/tt-metal#25143]
### Describe the bug

ttnn.experimental.all_gather_async seems not support 2D tensor

### Steps to reproduce the issue

```
import ttnn
import torch
def test_all_gather_async(shape, gather_dim):
    ttnn.set_fabric_config(ttnn.FabricConfig.FABRIC_1D)
    mesh_device = ttnn.open_mesh_device(ttnn.MeshShape(1, 8))
    try:
        input = torch.rand(shape, dtype=torch.float32)

        sharded_input = ttnn.from_torch(input, layout=ttnn.TILE_LAYOUT, device=mesh_device, mesh_mapper=ttnn.ShardTensorToMesh(mesh_device, dim=1))
        print("input shape : ", sharded_input.shape)

        compute_grid_size = mesh_device.compute_with_storage_grid_size()
        ccl_sub_device_crs = ttnn.CoreRangeSet(
            {ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(compute_grid_size.x - 1, compute_grid_size.y - 1))}
        )
        semaphore = ttnn.create_global_semaphore(mesh_device, ccl_sub_device_crs, 0)

        sharded_output_async = ttnn.experimental.all_gather_async(sharded_input, dim=gather_dim, 
        cluster_axis=1, mesh_device=mesh_device, topology=ttnn.Topology.Linear, multi_device_global_semaphore = semaphore)
        print("output shape : ", sharded_output_async.shape)

        output = ttnn.to_torch(sharded_output_async, device=mesh_device, mesh_composer=ttnn.ConcatMeshToTensor(mesh_device, dim=1))

    except Exception as e:
        print(f"An error occurred: {e}")
    finally:
        ttnn.close_mesh_device(mesh_device)
        ttnn.set_fabric_config(ttnn.FabricConfig.DISABLED)

test_all_gather_async((64, 512), 0)
```

LOG:
```
input shape :  Shape([64, 64])
2025-07-16 03:52:49.515 | critical |          Always | ShapeBase[] index out of range. 2 not in [-4, 2) (assert.hpp:107)
An error occurred: TT_THROW @ /localdev/hkwon/tt-mlir/third_party/tt-metal/src/tt-metal/tt_metal/common/shape_base.cpp:34: tt::exception
info:
ShapeBase[] index out of range. 2 not in [-4, 2)
```

**Comment by ntarafdar** [comment]
Use all_gather_async.

Example here https://github.com/tenstorrent/tt-metal/blob/main/tests/nightly/t3000/ccl/test_minimal_all_gather_async.py

**Comment by grechsteinerTT** [comment]
Other Notes:
- We currently only support 4D shapes, for the time being you will have to reshape your input shape to `[1, 1, x, y]` (and adjust your gather dim accordingly)
- A vector containing 2 semaphores is required

**Comment by amorrisonTT** [comment]
Implemented in #30998
